controller/controller.go

```diff
@@ -698,6 +704,96 @@ func (s *Service) heartbeatController(ctx context.Context, addr *url.URL) {
 	}
 }
 
+func (s *Service) watchModuleChanges(ctx context.Context, sendChange func(response *ftlv1.PullSchemaResponse) error) error {
```

Comment:
@alecthomas this seems like a lot of code to achieve what we want here :)

Any thoughts on how to refactor this a bit so it's not so much code? or does this look ok?

Reply:
You shouldn't need two loops. Get rid of the first one and just have the second one, that will diff against an empty map.

Reply:
Good point! I had a different implementation before which required it, but the refactor helped with that :) Will update.